### PR TITLE
Rfid support

### PR DESCRIPTION
**Pull request by Olverine**
# RFID
RFID tags can be used to start the station. This can be done either by storing the UID of the RFID tag in the EEPROM or by sending the UID to a MQTT topic so that the station can be started remotely.

## Prerequisites
In order to use RFID, a PN532 RFID scanner module is required. The module should be connected to the ESP32 board on pins 21 (SDA) and 22 (SCL). Make sure the module is set to use I2C for communication.

## Set up
- In the web gui, click the System tab
- Toggle the “Enabled” switch in the RFID box to on

The RFID module should now be activated. When the station is paused, a message should be displayed on the LCD every few seconds saying “Scan RFID tag to start”.

## RFID handling through MQTT
When the station detects an RFID tag, the unique UID of the tag will be sent to the MQTT topic: `<base-topic\>/rfid`. A subscriber can then authenticate the tag and start the station remotely.

## Offline RFID handling
If MQTT is unavailable, RFID tags can be authenticated offline. In order to do this, the UID of any authorized tags need to be stored in advance. 
### Storing an authorized tag
- In the web gui, click the System tab
- In the RFID box, click the “Scan” button under “Manage scanned tag”
- Place your RFID tag on the RFID scanner module
- When the UID of the tag shows up underneath the “Scan” button in the web gui, click the “Register tag” button.
### Removing an authorized tag
If the UID of the tag is known:
- In the web gui, click the System tab
- In the RFID box, find the UID in the list of registered tags
- Click “Remove” to the right of the UID you would like to remove

If the UID of the tag is unknown:
- In the web gui, click the System tab
- In the RFID box, click the “Scan” button under “Manage scanned tag”
- Place your RFID tag on the RFID scanner module
- When the UID of the tag shows up underneath the “Scan” button in the web gui, click the “Remove tag” button.

### CODE REVIEW

**Comment by jeremypoulter** [comment]
Hi, thanks for the PR, RFID is a long awaited feature. Unfortunately the timer code is a bit of a problem. I have been working on major changes in that area as part of #139. In theory this should remove the need for adding the timer code from this PR, it should just be able to sit on top of the new scheduler module. 

If you are willing please could you re-base your code on the jeremypoulter/issue5 branch? I appreciate that this is still pre-alpha code but would be great to include this in the new v4 firmware. Happy to help where needed.

**Comment by Olverine** [comment]
Thank you for reviewing our PR! I will re-base the branch and try to utilize the scheduler.

File: src/rfid.cpp [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+String getUidBytes(card NFCcard){
+    String bytes = "";
+    bytes.concat((char)NFCcard.uidlenght);
+    for(int i = 0; i < NFCcard.uidlenght; i+= 1){
+        bytes.concat((char)NFCcard.uid[i]);
+    }
+
+    Serial.println();
+    Serial.println(bytes);
+    Serial.println();
```
[the remark this answers is not in the document]
Comment (Olverine) [reply src/rfid.cpp:81]: This is a deprecated method that I forgot to clean up. Thank you for pointing it out!

File: src/rfid.cpp [hunk 2]
```diff
@@ -0,0 +1,199 @@
+#define STATUS_NOT_ENABLED 0
```
[the remark this answers is not in the document]
Comment (Olverine) [reply src/rfid.cpp:1]: Do you think it would be a good idea to make it into a `MicroTask`?

[the remark this answers is not in the document]
Comment (jeremypoulter) [reply src/rfid.cpp:1]: Yes please, but I haven't really written any documentation, so I wasn't going to insist on it.

File: src/rfid.cpp [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+    int schedulerDay = 1 << day;
+    int hour = offset / 3600;
+    int minute = (offset % 3600) / 60;
+    int second = offset % 60;
+    _scheduler->addEvent(0xFFFFFFFF, hour, minute, second, schedulerDay, EvseState(EvseState::Disabled));
+}
+
+void RfidTask::abortTimer(){
+    _scheduler->removeEvent(0xFFFFFFFF);
+}
```
Comment (Olverine) [inline src/rfid.cpp:109]: This does not seem to stop the event from getting triggered. Is this to be expected? How do I abort an event in the scheduler that has not yet been triggered?

File: src/rfid.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (Olverine) [reply src/rfid.cpp]: Would you like to take a look at the current implementation? Am I using MicroTask correctly?

[the remark this answers is not in the document]
Comment (jeremypoulter) [reply src/rfid.cpp]: I can not see anything obvious wrong from a quick scan, I will take a closer look. What issues are you seeing?

[the remark this answers is not in the document]
Comment (Olverine) [reply src/rfid.cpp]: No issues. I just wanted to make sure I understand the concept correctly

**Comment by jeremypoulter** [comment]
I have put I little bit more thought in this. I think this use case should be using the EVSE manager to claim/release control of the EVSE rather than using the scheduler. Unfortunately I haven't written the particular bit do do the time based charging :( 

Essentially you would do something like:

```c++
  // Lock the EVSE until RFID card presented
  EvseProperties props(EvseState::Disabled);
  evse.claim(EvseClient_OpenEVSE_RFID_Off, EvseManager_Priority_Lock, props);

  if(rfid_card_presented)
  {
    EvseProperties props(EvseState::Active);
    props.setTimeLimit(duration);
    evse.claim(EvseClient_OpenEVSE_RFID_On, EvseManager_Priority_Unlock, props);
  }
```

This would lock the EVSE by default until a card is presented and then allow charging for a given duration and then lock it again. The downside of doing it this way is that the priority would have to be higher than every other so would disable the schedule/PV divert... Maybe that is not a downside but a feature.... 

Anyway I will keep you updated.

**Comment by Olverine** [comment]
> This would lock the EVSE by default until a card is presented and then allow charging for a given duration and then lock it again. The downside of doing it this way is that the priority would have to be higher than every other so would disable the schedule/PV divert... Maybe that is not a downside but a feature....

We have been thinking about this issue as well. Since RFID is a form of access control, there should be no other way to activate the station as long as RFID is enabled. So we are thinking that it should have a higer priority than schedule/PV divert but perhaps there should be some kind of "authenticated state" that allows the scheduler an PV divert to activate the station when a tag has been authenticated. At the moment, scanning an RFID card will activate the station regardless of scheduler and PV divert settings which is probably a bad idea. It would be better if the station is activated when an RFID tag has been authenticated AND the scheduler is allowing the station to wake up. Do you agree?

I haven't looked into how this `evse.claim` method works very much but do you think it would be possible to keep the station disabled if it is unlocked by RFID but still locked by the scheduler and vice versa?

**Comment by jeremypoulter** [comment]
That is exactly the type of thing the claim is supposed to manage. If the RFID sets a high priority claim to disable the charger then lower level priorities can not override this. I am slowly moving everything over to the priority system so this should work well.

However from your description the example I gave above is slightly wrong, it you want the charger to work as it otherwise would then you need to release the claim for a period of time (or till the end of the charging session ?) then claim again. You could do this at the higher level for sure, but let me consider if there is a way to fit this in to the EVSE manager.

**Comment by jeremypoulter** [comment]
@Olverine I think this got closed because the issue5 branch was deleted after merge, please re-open targeting `master`

**Comment by Olverine** [comment]
> @Olverine I think this got closed because the issue5 branch was deleted after merge, please re-open targeting `master`

I can't find a way to re-open the PR. It might be a permission problem or it could be that the target branch no longer exists. It doesn't seem like I can change the target branch at this point however. Is it possible for you to re-open?

**Comment by jeremypoulter** [comment]
Sadly not :( @chris1howell are you able to re-open and change the target branch to `master`?

**Comment by chris1howell** [comment]
I will try.

File: src/rfid.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (Farnibull) [reply src/rfid.cpp]: Good morning,
I have been using the OpenEVSE terminal for a while with RFID. it works well.
I ordered the ESP32-gateway card for the Ethernet connection. Where should I connect my PN532 RFID to my ESP32-gateway on the ESP32-gateway? (SCL / SDA?
Thanks for your help

**Pull request closed without merging**